### PR TITLE
Implemented SA1129 analyzer, code fix, and tests

### LINKED ISSUE #1482: Implement SA1129 (DoNotUseDefaultValueTypeConstructor)
**Category:** Readability
**Title:** Use `default(T)` to create default value type instances
**Description:** A default instance of a value type `T` may be obtained by either `default(T)` or `new T()`. In these scenarios, the `default(T)` syntax should be used.

**Rationale:** This is especially applicable for code using `ImmutableArray<T>` (a value type). Code which formerly used `List<T>` or `T[]` (both reference types) need to be very careful not to mistake `new ImmutableArray<T>()` for the same behavior. This syntax results in a default `ImmutableArray<T>` instance, whose behavior more closely resembles `null` than any non-null instance (especially in regards to exceptions being thrown). Similar behavior can be seen for `Nullable<T>`.

:memo: This proposal only refers to the distinction between `default(T)` and `new T()`. Other default values, including `CancellationToken.None`, `0`, `0.0f`, `IntPtr.Zero`, and `RegexOptions.None`, would *not* produce a warning.

:memo: This proposal was inspired by code observed in [dotnet/Roslyn#5180 (comment)](https://github.com/dotnet/roslyn/pull/5180#discussion_r39443514).

### PR DESCRIPTION
**Pull request by vweijsters**
Fixes #1482

### CODE REVIEW

File: StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1129CodeFixProvider.cs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+        /// <inheritdoc/>
+        public override FixAllProvider GetFixAllProvider()
+        {
+            return FixAll.Instance;
+        }
+
+        public override Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            foreach (var diagnostic in context.Diagnostics.Where(d => this.FixableDiagnosticIds.Contains(d.Id)))
```
Comment (Maxwe11) [inline StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1129CodeFixProvider.cs]: There is no need in `Where` clause

**Comment by Maxwe11** [comment]
one minor comment but otherwise :+1:

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/ReadabilityRules/SA1129UnitTests.cs [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+        [Fact]
+        public async Task VerifyValueTypeWithParametersCreationAsync()
+        {
+            var testCode = @"public class TestClass
+{
+    public void TestMethod()
+    {
+        var v1 = new TestStruct(1);
+        var v2 = new TestStruct(1) { TestProperty = 2 };
+        var v3 = new TestStruct() { TestProperty = 2 };
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/ReadabilityRules/SA1129UnitTests.cs]: Nice catch finding this one!

:exclamation: You need to add one more for the following case (I tried adding it during the merge but it resulted in an AD0001 failure):

```csharp
var v4 = new TestStruct { TestProperty = 2 };
```

**Comment by sharwell** [comment]
I'm missing some tests:

### `struct` generic type constraint

The following code *should* produce a violation.

```csharp
public void Method<T>()
    where T : struct
{
    T value = new T(); // SA1129
}
```

### `new` generic type constraint

The following code *should not* produce a violation.

```csharp
public void Method<T>()
    where T : new()
{
    T value = new T();
}
```

**Comment by vweijsters** [comment]
Updated with CR feedback

**Pull request merged**